docs/docs/concepts/low_level.md

```diff
@@ -0,0 +1,427 @@
+
+interface State {
+  foo: number;
+  bar: string[];
+}
+
+const graphBuilder = new StateGraph<State>({
+  channels: {
+    foo: null,
+    bar: null,
+  }
+});
+```
+
+In this example, no reducer functions are specified for any key. Let's assume the input to the graph is `{ foo: 1, bar: ["hi"] }`. Let's then assume the first `Node` returns `{ foo: 2 }`. This is treated as an update to the state. Notice that the `Node` does not need to return the whole `State` schema - just an update. After applying this update, the `State` would then be `{ foo: 2, bar: ["hi"] }`. If the second node returns `{ bar: ["bye"] }` then the `State` would then be `{ foo: 2, bar: ["bye"] }`
+
+**Example B:**
+
+```typescript
+import { StateGraph } from "@langchain/langgraph";
+
+interface State {
+  foo: number;
+  bar: string[];
+}
+
+const graphBuilder = new StateGraph<State>({
+  channels: {
+    foo: null,
+    bar: {
+      reducer: (state: string[], update: string[]) => state.concat(update),
+      default: () => [],
+    },
+  }
+});
+```
+
+In this example, we've updated our `bar` field to be an object containing a `reducer` function. This function will always accept two positional arguments: `state` and `update`, with `state` representing the current state value, and `update` representing the update returned from a `Node`. Note that the first key remains unchanged. Let's assume the input to the graph is `{ foo: 1, bar: ["hi"] }`. Let's then assume the first `Node` returns `{ foo: 2 }`. This is treated as an update to the state. Notice that the `Node` does not need to return the whole `State` schema - just an update. After applying this update, the `State` would then be `{ foo: 2, bar: ["hi"] }`. If the second node returns`{ bar: ["bye"] }` then the `State` would then be `{ foo: 2, bar: ["hi", "bye"] }`. Notice here that the `bar` key is updated by concatenating the two arrays together.
+
+### MessageState
+
+`MessageState` is one of the few opinionated components in LangGraph. `MessageState` is a special state designed to make it easy to use a list of messages as a key in your state. Specifically, `MessageState` is defined as:
+
+```typescript
+import { BaseMessage } from "@langchain/langgraph";
+
+interface MessageState {
+  messages: BaseMessage[];
+}
+
+export class MessageGraph extends StateGraph {
+  constructor() {
+    super({
+      channels: {
+        __root__: {
+          reducer: messagesStateReducer,
+          default: () => [],
+        },
+      },
+    });
+  }
+}
+```
+
+What this is doing is creating a `State` with a single key `messages`. This is a list of `BaseMessage`s, with [`messagesStateReducer`](https://langchain-ai.github.io/langgraphjs/reference/functions/index.messagesStateReducer.html) as a reducer. `messagesStateReducer` basically adds messages to the existing list (it also does some nice extra things, like convert from OpenAI message format to the standard LangChain message format, handle updates based on message IDs, etc).
```

Comment:
Sorry this is bit confusing. 

Isn't BaseMessage imported from '@langchain/core/messages'?

It also reads like `MessageState` is supposed to imported and used? But it's not imported in the example.

The docs use the pattern below all over the place but this seems like it introduces a new implementation which is confusing.

I'm not sure if MessageState is WIP somewhere and I'm missing it. 

```ts
// this pattern is used throughout the docs
// Define the state interface
interface AgentState {
  messages: BaseMessage[];
}

// Define the graph state
const graphState: StateGraphArgs<AgentState>["channels"] = {
  messages: {
    reducer: (x: BaseMessage[], y: BaseMessage[]) => x.concat(y),
  },
};
```

How is the graph constructed after this?
```typescript
interface State extends MessagesState {
  documents: Array<string>
}
```

Again it seems like the expected pattern would be?
```ts
const graphState: StateGraphArgs<AgentState>["channels"] = {
  messages: {
    reducer: (x: BaseMessage[], y: BaseMessage[]) => x.concat(y),
  },
  documents: {
     reducer: `implementation here`
  },
};
```